src/undocommand.js

```diff
@@ -0,0 +1,200 @@
+			const updatedDeltas = this.editor.document.history.getTransformedDelta( undoDeltaReversed );
+
+			for ( let delta of updatedDeltas ) {
+				for ( let operation of delta.operations ) {
+					this.editor.document.applyOperation( operation );
+				}
+			}
+		}
+
+		// Get the selection state stored with this batch.
+		const selectionState = this._batchSelection.get( undoBatch );
+
+		// Take all selection ranges that were stored with undone batch.
+		const ranges = selectionState.ranges;
+
+		// The ranges will be transformed by deltas from history that took place
+		// after the selection got stored.
+		const deltas = this.editor.document.history.getDeltas( undoBatch.deltas[ 0 ].baseVersion ) || [];
+
+		// This will keep the transformed ranges.
+		const transformedRanges = [];
+
+		for ( let originalRange of ranges ) {
+			// We create `transformed` array. At the beginning it will have only the original range.
+			// During transformation the original range will change or even break into smaller ranges.
+			// After the range is broken into two ranges, we have to transform both of those ranges separately.
+			// For that reason, we keep all transformed ranges in one array and operate on it.
+			let transformed = [ originalRange ];
+
+			for ( let delta of deltas ) {
+				for ( let operation of delta.operations ) {
+					// We look through all operations from all deltas.
+
+					for ( let t = 0; t < transformed.length; t++ ) {
+						// We transform every range by every operation.
+						// We keep current state of transformation in `transformed` array and update it.
+						let result;
+
+						switch ( operation.type ) {
+							case 'insert':
+								result = transformed[ t ].getTransformedByInsertion( operation.position, operation.nodeList.length, true );
+								break;
+
+							case 'move':
+							case 'remove':
+							case 'reinsert':
+								result = transformed[ t ].getTransformedByMove( operation.sourcePosition, operation.targetPosition, operation.howMany, true );
+								break;
+						}
+
+						// If we have a transformation result, we substitute it in `transformed` array with
+						// the range that got transformed. Keep in mind that the result is an array
+						// and may contain multiple ranges.
+						if ( result ) {
+							transformed.splice( t, 1, ...result );
+
+							// Fix iterator.
+							t = t + result.length - 1;
+						}
+					}
+				}
+			}
+
+			// After `originalRange` got transformed, we have an array of ranges. Some of those
+			// ranges may be "touching" -- they can be next to each other and could be merged.
+			// Let's do this. First, we have to sort those ranges because they don't have to be
+			// in an order.
+			transformed.sort( ( a, b ) => a.start.isBefore( b.start ) ? -1 : 1 );
+
+			// Then we check if two consecutive ranges are touching. We can do it pair by pair
+			// in one dimensional loop because ranges are sorted.
+			for ( let i = 1 ; i < transformed.length; i++ ) {
```

Comment:
I just wanted to leave some comment in the middle of this function, so it looks like we've read and understood that code.